### PR TITLE
Decouple flowtip-react-dom rendering logic

### PR DESCRIPTION
**Pull request by 10xjs**
Remove the component rendering and style generation logic from within the `FlowTip` react component and allow it to be pluggable through a `render` interface.

The previous behavior is preserved through the default configuration. Different renderer types (other than a `Content` and `Tail` component interface) can be easily replace the default renderer.

### Consumer-facing Changes:

Previously, the appearance could only be customized through the `tail` and `content` components. **This interface remains supported and unchanged.**

```jsx
import FlowTip from 'flowtip-react-dom';

  <FlowTip target={this.state.target} content={Content} tail={Tail}>
    FlowTip Content
  </FlowTip>
```

Using the `render` prop, the consumer has lower-level access to the rendering behavior. This pattern has far fewer assumptions about the style or shape of the rendered DOM nodes. 

```jsx
import FlowTip from 'flowtip-react-dom';

  <FlowTip
    target={this.state.target}
    render={({onContentSize, onTailSize, state, props}) => {
      // entirely custom render logic
    }}
  >
    FlowTip Content
  </FlowTip>
```

### CODE REVIEW

File: packages/flowtip-core/src/Rect.js [hunk 1]
```diff
@@ -104,6 +104,16 @@ class Rect {
     );
   }
 
+  /**
+   * Determne if a rect-like object has valid positive area.
```
Comment (izaakschroeder) [inline packages/flowtip-core/src/Rect.js]: Spelling 😄

File: packages/flowtip-react-dom/src/defaultRender.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            result={state.result}
+          >
+            <ResizeObserver onResize={onTailSize} />
+          </TailComponent>
+        )}
+      </ContentComponent>
+    );
+  }
+
+  return null;
```
Comment (izaakschroeder) [inline packages/flowtip-react-dom/src/defaultRender.js]: Should there be a warning in this case? Or use of `invariant` instead?

File: packages/flowtip-react-dom/src/util/render.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * Create a CSS border triangle facing a given region.
+ *
+ * @param   {string} region -`top`, `right`, `bottom`, or `left`.
+ * @param   {string} color - CSS color string.
+ * @param   {number} size - Height of the triangle in px.
+ * @returns {Object} Border triangle style.
+ */
+export const cssTriangle = (region: Region, color: string, size: number) => {
```
Comment (izaakschroeder) [inline packages/flowtip-react-dom/src/util/render.js]: What is this used for? There's no reference to it anywhere. Is this intended to be part of a separate thing to help consumers quickly generate carets for use in a tail component?

File: packages/flowtip-react-dom/src/util/state.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    // the tail was rendered using the static dummy result that uses the
+    // bottom region
+    return state.tail.width;
+  }
+
+  return 0;
+}
+
+/**
+ * Get current minimum linear overlap value.
```
Comment (izaakschroeder) [inline packages/flowtip-react-dom/src/util/state.js:92]: I know you've used this term before but is there a definition of what "linear" overlap is vs some other kind of overlap? This comment has little to do with the PR itself, just if it's going to be part of the nomenclature for this library a working definition would be handy. A cursory google reveals no common definition or use of this term.

File: packages/flowtip-core/src/Rect.js [hunk 5]
```diff
@@ -104,6 +104,16 @@ class Rect {
...
+   *
+   * @param   {object} [rect] A rect-like object.
```
Comment (izaakschroeder) [inline packages/flowtip-core/src/Rect.js]: Is there a reason this `object` is lowercase and all the other ones are not?

**Review by izaakschroeder — APPROVED** [review]
The ideas behind this seem pretty great to me. Left a few thoughts. The only thing that might be an area to dry up is there seems to be some repeat logic around `document.documentElement`. Nothing critical though.

File: packages/flowtip-react-dom/src/util/state.js (hunk 4 above)
Comment (10xjs) [reply packages/flowtip-react-dom/src/util/state.js:92]: `overlap` is one of the named config options of `flowtip-core`

Comment (izaakschroeder) [reply packages/flowtip-react-dom/src/util/state.js:92]: No concern with `overlap` – as above more curious what the `linear` qualifier means.

Comment (10xjs) [reply packages/flowtip-react-dom/src/util/state.js:92]: "Linear" here refers to parallel with the edge of the target facing the content, as opposed to perpendicular.

The overlap related computations are effectively 2D with respect to one of the four regions.

**Pull request merged**